### PR TITLE
Optimize revision recording, speed up update dashboard API by 25% 🚀

### LINKED ISSUE #34721: Update dashboard performance 48
We've noticed the performance of updating a dashboard has been much slower recently, it took ~3 seconds to save a dashboard on our stats([slack thread](https://metaboat.slack.com/archives/C01MS7DQKR6/p1697454979662699))

This issue is used to track and report findings.

### PR DESCRIPTION
**Pull request by qnkhuat**
An attempt to resolve https://github.com/metabase/metabase/issues/34721 by reducing the number of DB calls when recording revision.

Other changes:
- adding schema for event system in "metabase.events.schema"
- decouple the info in the `event` of `(events/publish-event! topic event)`, before it's used to store everything it's needed for the event to process, it's usually a DB object like Card/Dashboard, and assoced with `actor-id`, `dashcards`... changing this to `{:object dashboard, :other-info 1}` to makes it explicit what's needed. also we can reuse the `object` to record revision so we don't have to refetch it.

Benchmarking ([script](https://gist.github.com/qnkhuat/65d63319fa3d63e52b08c1afe17094f4)):

master
- `GET /api/dashboard/:id` reports 16 DB calls
  -  takes ~14.14ms (N=1000)
- `POST /api/dashboard` reports 23 DB calls
  - takes ~14.7ms (N= 1000)
- `PUT /api/dashboard/:id` reports 22 DB calls
  - takes ~16.19ms (N=1000)
- `PUT /api/dashboard/cards` report 50 DB calls
  - takes ~ 34.6ms (N=1000)

this branch
- `GET /api/dashboard/:id` reports 16 DB calls
  -  takes ~12ms (N=1000)
- `POST /api/dashboard/:id` reports 20 DB calls
  - takes ~11.4ms (N= 1000)
- `PUT /api/dashboard/:id` reports 25 DB calls
  - takes ~14.2ms (N=1000)
- `PUT /api/dashboard/cards` report 39 DB calls
  - takes ~ 25.4ms (N=1000)

### CODE REVIEW

File: src/metabase/api/card.clj [hunk 1]
```diff
@@ -608,20 +608,6 @@ saved later when it is ready."
     (validation/check-embedding-enabled)
     (api/check-superuser)))
 
-(defn- publish-card-update!
```
Comment (qnkhuat) [inline src/metabase/api/card.clj:611]: there are no subscribers for card-unarchive and card-archive, also it has never been published before because archived and `(:card archive)` is identical when it's called this function.

so I've removed it

File: src/metabase/events/revision.clj [hunk 2]
```diff
@@ -12,25 +10,28 @@
...
+                                       {:mode model :event event})))
+           user-id (or (:actor-id event)
+                       (throw (ex-info "Event does not have ID associated with it"
+                                       {:mode model :event event})))]
+       (when-not (t2/instance-of? model object)
+         (throw (ex-info "object must be a model instance" {:object object :model model})))
+       (revision/push-revision! {:entity       model
                                  :id           id
-                                 :object       (api/check-404 (t2/select-one model :id id))
+                                 :object       object
```
Comment (qnkhuat) [inline src/metabase/events/revision.clj:29]: optimization: no need to select the object again

File: src/metabase/models/dashboard.clj [hunk 3]
```diff
@@ -216,11 +216,19 @@
 
 (defmethod revision/serialize-instance :model/Dashboard
   [_model _id dashboard]
-  (-> (apply dissoc dashboard excluded-columns-for-dashboard-revision)
-      (assoc :cards (vec (for [dashboard-card (dashcards dashboard)]
-                           (-> (apply dissoc dashboard-card excluded-columns-for-dashcard-revision)
-                               (assoc :series (mapv :id (dashboard-card/series dashboard-card)))))))
-      (assoc :tabs (map #(apply dissoc % excluded-columns-for-dashboard-tab-revision) (tabs dashboard)))))
+  (let [cards (or (:cards dashboard)
```
Comment (qnkhuat) [inline src/metabase/models/dashboard.clj:219]: optimization: only hydrates when needed

File: src/metabase/models/dashboard_card.clj [hunk 4]
```diff
@@ -89,16 +88,23 @@
 
 ;;; --------------------------------------------------- HYDRATION ----------------------------------------------------
 
-(mi/define-simple-hydration-method series
+(mi/define-batched-hydration-method series
```
Comment (qnkhuat) [inline src/metabase/models/dashboard_card.clj:91]: series has always been batch hydrated, so making it more efficient

File: src/metabase/models/dashboard_card.clj [hunk 5]
```diff
@@ -111,8 +117,9 @@
 (mu/defn retrieve-dashboard-card
```
Comment (qnkhuat) [inline src/metabase/models/dashboard_card.clj:117]: this is currently only used in test tho.

File: src/metabase/events/activity_feed.clj [hunk 6]
```diff
@@ -47,99 +50,120 @@
...
+        create-delete-details
+        #(select-keys (:object %) [:description :name])
 
         add-remove-card-details
-        (fn [{:keys [dashcards] :as obj}]
-          ;; we expect that the object has just a dashboard :id at the top level
-          ;; plus a `:dashcards` attribute which is a vector of the cards added/removed
-          (-> (t2/select-one [Dashboard :description :name], :id (events/object->model-id topic obj))
-              (assoc :dashcards (for [{:keys [id card_id]} dashcards]
-                                  (-> (t2/select-one [Card :name :description], :id card_id)
```
Comment (qnkhuat) [inline src/metabase/events/activity_feed.clj:60]: N+1 with the number of dash cards 🙀

File: src/metabase/events/activity_feed.clj [hunk 7]
```diff
@@ -47,99 +50,120 @@
...
 (derive ::metric-event ::event)
 (derive :event/metric-create ::metric-event)
 (derive :event/metric-update ::metric-event)
 (derive :event/metric-delete ::metric-event)
 
 (methodical/defmethod events/publish-event! ::metric-event
-  [topic object]
-  (let [details-fn  #(select-keys % [:name :description :revision_message])
-        table-id    (:table_id object)
+  [topic event]
```
Comment (calherries) [inline src/metabase/events/activity_feed.clj:89]: Again, I think it would be nicer to destructure all the keys of event here

File: src/metabase/models/revision.clj [hunk 8]
```diff
@@ -153,37 +154,37 @@
       ;; Filter out irrelevant info
       (dissoc :model :model_id :user_id :object)))
 
-(defn revisions
+(mu/defn revisions
   "Get the revisions for `model` with `id` in reverse chronological order."
-  [model id]
+  [model :- [:fn mdb.u/toucan-model?]
+   id    :- pos-int?]
   {:pre [(mdb.u/toucan-model? model) (integer? id)]}
```
Comment (calherries) [inline src/metabase/models/revision.clj:161]: You don't need to duplicate the schema with the `:pre` block
```suggestion
```

**Review by calherries — APPROVED** [review]
Approving because it works and my suggestions are all stylistic. I do want `:actor-id` renamed to `:user-id` everywhere though.

My biggest suggestion would be to make smaller PRs if you can help it next time. There's far too much stuff happening in this PR to review quickly and well.

File: src/metabase/events/schema.clj [hunk 9]
```diff
@@ -0,0 +1,133 @@
+(ns metabase.events.schema
```
Comment (calherries) [inline src/metabase/events/schema.clj:1]: For a follow-up PR: Rather than centralize event schemas in one place and have all this code to merge them together, I suggest dropping the schemas and thinking about how to create a multimethod like `validate-event` that is dispatched by event topic. 

Imagine if all our functions has schemas in one namespace! We'd waste a lot of our dev time just switching between files and following references

File: src/metabase/events/schema.clj [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+(def ^:private dashboard-events-schemas
+  {:event/dashboard-read             dashboard-default
+   :event/dashboard-create           dashboard-default
+   :event/dashboard-update           dashboard-default
+   :event/dashboard-reposition-cards dashboard-update-with-dashcards
+   :event/dashboard-remove-cards     dashboard-update-with-dashcards
+   :event/dashboard-add-cards        dashboard-update-with-dashcards
+   :event/dashboard-add-tabs         dashboard-update-with-tab-ids
+   :event/dashboard-update-tabs      dashboard-update-with-tab-ids
+   :event/dashboard-remove-tabs      dashboard-update-with-tab-ids})
```
Comment (calherries) [inline src/metabase/events/schema.clj:37]: This could be much briefer
```clojure
(let [default-schema [:map {:closed true}
                      [:actor-id pos-int?]
                      [:object [:fn #(t2/instance-of? :model/Dashboard %)]]]
      with-dashcards (mut/assoc default-schema :dashcards [:sequential [:map [:id pos-int?]]])
      with-tabs      (mut/assoc default-schema :tab-ids [:sequential pos-int?])]
  (def ^:private dashboard-events-schemas
    {:event/dashboard-read             dashboard-default
     :event/dashboard-create           dashboard-default
     :event/dashboard-update           dashboard-default
     :event/dashboard-reposition-cards dashboard-update-with-dashcards
     :event/dashboard-remove-cards     dashboard-update-with-dashcards
     :event/dashboard-add-cards        dashboard-update-with-dashcards
     :event/dashboard-add-tabs         dashboard-update-with-tab-ids
     :event/dashboard-update-tabs      dashboard-update-with-tab-ids
     :event/dashboard-remove-tabs      dashboard-update-with-tab-ids}))
```

File: src/metabase/events.clj [hunk 11]
```diff
@@ -107,41 +108,32 @@
...
+     (log/debugf "Publishing %s event:\n\n%s" (u/colorize :yellow (pr-str topic)) (u/pprint-to-str event))
+     (assert (and (qualified-keyword? topic)
+                  (isa? topic :metabase/event))
+             (format "Invalid event topic %s: events must derive from :metabase/event" (pr-str topic)))
+     (assert (map? event)
+             (format "Invalid event %s: event must be a map." (pr-str event)))
+     (try
+      (if-let [schema (events.schema/topic->schema topic)]
+        (mu/validate-throw schema event)
+        (log/warn (format "Missing schema for topic: %s" topic)))
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/events.clj:123]: on schema living with the code: the thing with the event system is that one topic key can have multiple handlers, 

take `:event/card-create` for example, there are 4 subscribers defined in 4 namespaces.

So defining a schema on defmethod will force us to duplicate the schema everywhere.

but that is not the biggest problem tho, I think we can have a single ns for schema and reuse it across event namespaces. The biggest problem right now is methodical haven't supported it https://github.com/camsaul/methodical/issues/145

[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/events.clj:123]: > Also I wouldn't warn if a schema is missing for a topic

yea I wanted it to warn to aid me in adding events with missing schema. now it's done so I'll remove the log.

File: src/metabase/events/schema.clj (hunk 9 above)
Comment (qnkhuat) [reply src/metabase/events/schema.clj:1]: see my other [comment](https://github.com/metabase/metabase/pull/34947/files#r1375607002) for details on why I chose this approach.

File: src/metabase/models/pulse.clj [hunk 12]
```diff
@@ -507,9 +506,7 @@
                 [:collection_position {:optional true} [:maybe ms/PositiveInt]]
                 [:dashboard_id        {:optional true} [:maybe ms/PositiveInt]]
                 [:parameters          {:optional true} [:maybe [:sequential :map]]]]]
-  (let [pulse-id (create-notification-and-add-cards-and-channels! kvs cards channels)]
-    ;; return the full Pulse (and record our create event)
-    (events/publish-event! :event/pulse-create (retrieve-pulse pulse-id))))
+  (retrieve-pulse (create-notification-and-add-cards-and-channels! kvs cards channels)))
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/models/pulse.clj:509]: I've moved the publish-event into the api namespace.

File: src/metabase/models/dashboard.clj (hunk 3 above)
Comment (qnkhuat) [reply src/metabase/models/dashboard.clj]: renamed to dashcards https://github.com/metabase/metabase/pull/34947/commits/d542853d655500e2e2288514f83a1d6ec589179f

File: src/metabase/events/activity_feed.clj [hunk 13]
```diff
@@ -21,24 +20,28 @@
...
-                       ;; right now datasets are all models. In the future this will change so lets keep a breadcumb
-                       ;; around
-                       dataset? (assoc :original-model "card"))
-        query       (when (seq query)
-                      (try (qp/preprocess query)
-                           (catch Throwable e
-                             (log/error e (tru "Error preprocessing query:")))))
-        database-id (some-> query :database u/the-id)
-        table-id    (mbql.u/query->source-table-id query)]
+  [topic event]
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/events/activity_feed.clj]: https://github.com/metabase/metabase/pull/34947/commits/d542853d655500e2e2288514f83a1d6ec589179f

File: src/metabase/events/activity_feed.clj [hunk 14]
```diff
@@ -47,99 +50,120 @@
...
+     {:topic    topic
+      :model    "dashboard"
+      :model-id (:id dashboard)
+      :user-id  (:actor-id event)
+      :object   dashboard
+      :details  (case topic
+                  :event/dashboard-create       (create-delete-details event)
+                  :event/dashboard-delete       (create-delete-details event)
+                  :event/dashboard-add-cards    (add-remove-card-details event)
+                  :event/dashboard-remove-cards (add-remove-card-details event))})))
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/events/activity_feed.clj]: cool https://github.com/metabase/metabase/pull/34947/commits/d542853d655500e2e2288514f83a1d6ec589179f

File: src/metabase/events/driver_notifications.clj [hunk 15]
```diff
@@ -15,10 +15,11 @@
...
-  [topic database]
+  [topic event]
   ;; try/catch here to prevent individual topic processing exceptions from bubbling up.  better to handle them here.
   (try
-    ;; notify the appropriate driver about the updated database
-    (driver/notify-database-updated (:engine database) database)
-    (catch Throwable e
-      (log/warnf e "Failed to process driver notifications event. %s" topic))))
+   ;; notify the appropriate driver about the updated database
+   (let [database (:object event)]
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/events/driver_notifications.clj]: https://github.com/metabase/metabase/pull/34947/commits/d542853d655500e2e2288514f83a1d6ec589179f

File: src/metabase/events.clj (hunk 11 above)
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/events.clj]: https://github.com/metabase/metabase/pull/34947/commits/d542853d655500e2e2288514f83a1d6ec589179f

File: src/metabase/models/revision.clj [hunk 16]
```diff
@@ -153,37 +154,37 @@
...
-      :or {id (:id object), is-creation? false}}]
-  ;; TODO - rewrite this to use a schema
-  {:pre [(mdb.u/toucan-model? entity)
-         (integer? user-id)
-         (t2/exists? User :id user-id)
-         (integer? id)
-         (t2/exists? entity :id id)
-         (map? object)]}
-  (let [serialized-object (serialize-instance entity id (dissoc object :message))
+  [info :- [:map {:closed true}
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/models/revision.clj]: https://github.com/metabase/metabase/pull/34947/commits/d542853d655500e2e2288514f83a1d6ec589179f

File: src/metabase/models/dashboard_card.clj [hunk 17]
```diff
@@ -111,8 +117,9 @@
...
   "Fetch a single DashboardCard by its ID value."
   [id :- ms/PositiveInt]
-  (-> (t2/select-one :model/DashboardCard :id id)
-      (t2/hydrate :series)))
+  (-> (t2/select :model/DashboardCard :id id)
+      (t2/hydrate :series)
+      first))
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/models/dashboard_card.clj]: https://github.com/metabase/metabase/pull/34947/commits/d542853d655500e2e2288514f83a1d6ec589179f

File: test/metabase/api/dashboard_test.clj [hunk 18]
```diff
@@ -1445,7 +1445,7 @@
 
 (defn- dashcard-like-response
   [id]
-  (t2/hydrate (t2/select-one DashboardCard :id id) :series))
+  (first (t2/hydrate (t2/select DashboardCard :id id) :series)))
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply test/metabase/api/dashboard_test.clj]: https://github.com/metabase/metabase/pull/34947/commits/d542853d655500e2e2288514f83a1d6ec589179f

File: src/metabase/events/activity_feed.clj (hunk 7 above)
Comment (qnkhuat) [reply src/metabase/events/activity_feed.clj]: https://github.com/metabase/metabase/pull/34947/commits/d542853d655500e2e2288514f83a1d6ec589179f

File: src/metabase/events/schema.clj (hunk 10 above)
Comment (qnkhuat) [reply src/metabase/events/schema.clj]: https://github.com/metabase/metabase/pull/34947/commits/d542853d655500e2e2288514f83a1d6ec589179f

File: src/metabase/events/revision.clj [hunk 19]
```diff
@@ -12,25 +10,28 @@
...
-      (let [id      (or (:id event)
-                        (get event (keyword (str (u/lower-case-en (name model)) "_id")))
-                        (throw (ex-info "Event does not have ID associated with it"
-                                        {:mode model, :event event})))
-            user-id (events/object->user-id event)]
-        (revision/push-revision! :entity       model
+     (let [id      (or (:id object)
+                       (throw (ex-info "Event does not have ID associated with it"
+                                       {:mode model :event event})))
+           user-id (or (:actor-id event)
```
[the remark this answers is not in the document]
Comment (qnkhuat) [reply src/metabase/events/revision.clj]: did this in 0fa9b07033

File: src/metabase/query_processor/middleware/process_userland_query.clj [hunk 20]
```diff
@@ -92,12 +92,12 @@
...
-                                                   :cached       (:cached acc)
-                                                   :context      (:context execution-info)
-                                                   :ignore_cache (get-in execution-info
-                                                                         [:json_query :middleware :ignore-cached-results?])}))
+         (events/publish-event! :event/card-query {:user-id (:executor_id execution-info)
+                                                   :object  {:id           (:card_id execution-info)
+                                                             :cached       (:cached acc)
+                                                             :context      (:context execution-info)
+                                                             :ignore_cache (get-in execution-info
+                                                                                   [:json_query :middleware :ignore-cached-results?])}}))
```
Comment (calherries) [inline src/metabase/query_processor/middleware/process_userland_query.clj:100]: Why did you move this data under an `:object` key? It doesn't look like an object to me, just a bunch of data

**Review by calherries — CHANGES_REQUESTED** [review]

File: src/metabase/query_processor/middleware/process_userland_query.clj (hunk 20 above)
Comment (qnkhuat) [reply src/metabase/query_processor/middleware/process_userland_query.clj:100]: https://github.com/metabase/metabase/blob/7eb27d08d973ba987657d41316e8a48b6090a75b/src/metabase/events/view_log.clj#L81

this is the code that consumes this event and it's currently used to process `:event/dashboard-read`, `:event/table-read`, `:event/card-read` as well.

so I make it an object in order to not have to modify this defmethod.

another option is to create a new defmethod just to handle `:event/card-query`, is that better?

Comment (calherries) [reply src/metabase/query_processor/middleware/process_userland_query.clj:100]: Yes, I think creating a new defmethod to handle `:event/card-query` is a better option. We should avoid reusing names of keys inconsistently. The current handler for `:metabase.events.view-log/event` is effectively handling two different formats of data - one for actual views and one for queries.

File: src/metabase/events/schema.clj (hunk 9 above)
Comment (calherries) [reply src/metabase/events/schema.clj:1]: Still, it feels desirable to change `events.schema/topic->schema` to be a multimethod so you don't have to merge all the schemas into a map and you can define them in any namespace.

**Review by calherries — APPROVED** [review]
Looks good, thanks

File: src/metabase/models/revision/diff.clj [hunk 21]
```diff
@@ -119,14 +119,13 @@
...
-                             (diff-string k (k before) (k after)
-                                          (if (zero? i) (deferred-tru "this {0}" model-name) (deferred-tru "it"))))
-                           ks)))))
-:used-underscored-binding
-
-(let [x 1 y 2]
-  (clojure.core.match/match [:name x y]
-    [:name _ _]
-    "Ok"))
+      (loop [ks               ks
```
Comment (qnkhuat) [inline src/metabase/models/revision/diff.clj:122]: we had a failing e2e test because of this https://github.com/metabase/metabase/actions/runs/6714770177/job/18249901919?pr=34947
